### PR TITLE
Added Inverse Wishart distribution

### PR DESCRIPTION
**Pull request by giulio-zhou**

### CODE REVIEW

File: src/random/InvWishart.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  arma::mat scale;
+  int freeDeg;
+  double multivarGamma(int p, double x);
+  double logMultivarGamma(int p, double x);
+  int d;
+  double dimFactor, logDimFactor, normConst, logNormConst;
+  MultivarGaussian mvg;
+  Gamma g;
+  bool is_gen_ok, is_likeli_ok, is_loglikeli_ok;
+  const double logPI;
```
Comment (jxwuyi) [inline src/random/InvWishart.h]: I recommend add logPi to swiftDistribution.h

File: src/random/InvWishart.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  logNormConst = std::log(arma::det(scale)) * 0.5 + logDimFactor;
+
+  retvec.set_size(d, d);
+  mvg.init(arma::zeros<arma::mat>(d, 1), arma::inv_sympd(scale));
+  g.init(d);
+}
+
+double InvWishart::likeli(const arma::mat& x) {
+  if (!is_likeli_ok) {
+    g.init(d);
```
Comment (jxwuyi) [inline src/random/InvWishart.cpp]: this line should be moved to the constructer

File: src/random/InvWishart.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    is_likeli_ok = true;
+  }
+
+  return normConst * pow(arma::det(x), -(freeDeg + d + 1) * 0.5) *
+         exp(-0.5 * arma::trace(scale * arma::inv_sympd(x)));
+}
+
+double InvWishart::loglikeli(const arma::mat& x) {
+  if (!is_loglikeli_ok) {
+    g.init(d);
```
Comment (jxwuyi) [inline src/random/InvWishart.cpp]: also this

File: src/printer/CPPPrinter.cpp [hunk 4]
```diff
@@ -242,6 +242,7 @@ void CPPPrinter::print(code::Code* prog) {
   fprintf(file, "#include \"random/Geometric.h\"\n");
   fprintf(file, "#include \"random/Poisson.h\"\n");
   fprintf(file, "#include \"random/InvGamma.h\"\n");
+  fprintf(file, "#include \"random/InvWishart.h\"\n");
```
Comment (jxwuyi) [inline src/printer/CPPPrinter.cpp]: This line should be moved to line 270. These are all the distributions that do not need matrix.

**Comment by giulio-zhou** [comment]
Looking through the changes now. If you have a moment, wanted to ask about the use of `is_likeli_ok` and `is_gen_ok`. I'm actually not too clear on why we don't just do all of the initialization in the constructor.

File: src/random/InvWishart.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  is_gen_ok = false;
+  is_likeli_ok = false;
+  is_loglikeli_ok = false;
+
+  // Set constants
+  dimFactor = 1 / (pow(2, freeDeg * d * 0.5) * multivarGamma(d, freeDeg * 0.5));
+  logDimFactor = -log(2) * freeDeg * d * 0.5 -
+                 logMultivarGamma(d, freeDeg * 0.5);
+  normConst = pow(arma::det(scale), freeDeg * 0.5) * dimFactor;
```
Comment (giulio-zhou) [inline src/random/InvWishart.cpp]: will do.

File: src/random/InvWishart.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  logNormConst = std::log(arma::det(scale)) * 0.5 + logDimFactor;
```
Comment (giulio-zhou) [inline src/random/InvWishart.cpp]: will do.

File: src/random/InvWishart.cpp (hunk 2 above)
Comment (giulio-zhou) [inline src/random/InvWishart.cpp]: will do.

File: src/random/InvWishart.h (hunk 1 above)
Comment (giulio-zhou) [inline src/random/InvWishart.h]: sounds good

File: src/printer/CPPPrinter.cpp (hunk 4 above)
Comment (giulio-zhou) [inline src/printer/CPPPrinter.cpp]: done.

**Comment by giulio-zhou** [comment]
I'll briefly describe my basic test for Inverse Wishart. Given scale matrix $$\Psi$$ and $$\nu$$ as the number of degrees of freedom, the mean of the distribution is $$\frac{\Psi}{\nu - p - 1}$$. In this case, I choose $$\nu$$ to be 4 so that we essentially get back the original scale matrix in expectation. Running this both in Java Blog and Swift, I've that the average matrices returned are very close to the input scale matrix.

**Comment by jxwuyi** [comment]
For the "is_likeli" stuff.

Yes, for correctness, it is true. However, for efficiency, having a flag "is_likeli_ok" is better.
Think about this BLOG program:

X ~ UniformReal(-1, 1)
Y ~ UniformReal(-10,X)

In this case, X is a random variable, so in the compiled code, the UniformReal class for Y cannot be initialized statically. Hence, in each sampling iteration, we need to call the init() method of UniformReal distribution.

Note that the init for calling loglikelihood for UniformReal is very expensive: including two calls of function log() (you can compare the running time of a=log(a) against a*=1.5 and a+=1.5). When the user is only using the likelihood instead of log-like, we do not need to execute those very expensive log() calls.

This is the reason, which holds for so many distributions. This becomes serious when the model is very small. Definitely, if the model itself is very complicated, or all distributions can be statically initialized, it does not matter.

File: src/random/InvWishart.cpp [hunk 7]
```diff
@@ -37,17 +36,13 @@ void InvWishart::init(const arma::mat& _scale, int _freeDeg) {
   dimFactor = 1 / (pow(2, freeDeg * d * 0.5) * multivarGamma(d, freeDeg * 0.5));
   logDimFactor = -log(2) * freeDeg * d * 0.5 -
```
Comment (jxwuyi) [inline src/random/InvWishart.cpp]: @giulio-zhou Seems that you forgot to move these two lines

**Comment by giulio-zhou** [comment]
Delete a few statements from method inits and added a test. 
[was about to send this] Should I leave the dimFactor and logDimFactor constants as they are?

To comment on the test, the input cov matrix is what the query should return in expectation. I've verified that the same result is obtained on Java BLOG; the Java implementation returns a list of covariance matrices followed by their likelihoods. To compute the average returned covariance matrix, I  took the weighted sum of matrices, weighted by their likelihoods.

File: src/random/InvWishart.cpp (hunk 7 above)
Comment (giulio-zhou) [inline src/random/InvWishart.cpp]: Yeah, I figured those should be removed as well.

File: src/random/InvWishart.cpp [hunk 8]
```diff
@@ -32,11 +32,7 @@ void InvWishart::init(const arma::mat& _scale, int _freeDeg) {
   is_likeli_ok = false;
   is_loglikeli_ok = false;
 
-  // Set constants
-  dimFactor = 1 / (pow(2, freeDeg * d * 0.5) * multivarGamma(d, freeDeg * 0.5));
-  logDimFactor = -log(2) * freeDeg * d * 0.5 -
-                 logMultivarGamma(d, freeDeg * 0.5);
-
+  // Initialize distributions
   mvg.init(arma::zeros<arma::mat>(d, 1), arma::inv_sympd(scale));
```
Comment (jxwuyi) [inline src/random/InvWishart.cpp]: better move to line 65

When just computing likelihood, not necessary to compute an expensive inv_symd

Comment (giulio-zhou) [inline src/random/InvWishart.cpp]: Yeah, makes sense. All we want to do here is initialize distributions which are absolutely necessary. The Gamma distribution shouldn't be too expensive to evaluate, right?

Comment (jxwuyi) [inline src/random/InvWishart.cpp]: I thought about this. Gamma should be fine.

**Comment by jxwuyi** [comment]
Btw, I think this can be directly push to master

**Comment by giulio-zhou** [comment]
Ok, sounds good.

**Comment by jxwuyi** [comment]
@giulio-zhou  I accidentally closed the branch. Do you know how to re-open it?

**Comment by giulio-zhou** [comment]
reopened the PR

**Comment by jxwuyi** [comment]
looks good

**Pull request merged**